File: Episode021/WebBrokerDistroLessDemo/CodeExamplesU.pas
{
  This unit contains code examples used in the WebStencils demo.
  It provides a centralized location for all code snippets displayed in the demo pages.
}

unit CodeExamplesU;

interface

uses
  System.Generics.Collections, System.SysUtils, Web.Stencils;

type

  TCodeExample = class
  private
    FId: string;
    FCode: string;
  public
    constructor Create(AId: string; const ACode: string);
    property Id: string read FId;
    property Code: string read FCode write FCode;
  end;

{
  TCodeExamples manages a collection of code examples used throughout the demo.
  It uses a TObjectDictionary to store and retrieve examples efficiently.
}
  TCodeExamples = class
    FItems: TObjectDictionary<string, TCodeExample>;
    FWebStencilsEngine: TWebStencilsEngine;
  private
    procedure InitValues;
  public
    procedure Add(AId: string; const ACode: string);
    function Find(AId: string): TCodeExample;
    constructor Create(AWebStencilsEngine: TWebStencilsEngine);
    destructor Destroy; override;
  end;

implementation

{ TCodeExample }

constructor TCodeExample.Create(AId: string; const ACode: string);
begin
  FId := AId;
  FCode := ACode;
end;

{ TCodeExamples }

constructor TCodeExamples.Create(AWebStencilsEngine: TWebStencilsEngine);
begin
  inherited Create;
  FItems := TObjectDictionary<string, TCodeExample>.Create;
  FWebStencilsEngine := AWebStencilsEngine;
  self.InitValues;
  for var Pair in FItems do
  begin
    var Key := Pair.Key;
    var CodeExample := Pair.Value;
    FWebStencilsEngine.AddVar(Key, CodeExample);
  end;
end;

destructor TCodeExamples.Destroy;
begin
  FItems.Free;
  inherited;
end;

procedure TCodeExamples.Add(AId: string; const ACode: string);
begin
  FItems.Add(AId, TCodeExample.Create(AId, ACode));
end;

function TCodeExamples.Find(AId: string): TCodeExample;
begin
  FItems.TryGetValue(AId, Result);
end;

procedure TCodeExamples.InitValues;
begin
  Add('codeAtSymbol', '@object.value');
  Add('codeDotNotation',
      '''
      <h2>User Profile</h2>
      <p>Name: @user.name</p>
      <p>Email: @user.email</p>
      ''');
  Add('codeComments',
      '''
      @* This is a comment and will not appear in the output *@
      <p>This will appear in the output</p>
      ''');
  Add('codeIfAndElse',
      '''
      @if user.isLoggedIn {
          <p>Welcome, @user.name!</p>
      }
      @else {
          <p>Please log in to continue.</p>
      }
      ''');
  Add('codeIfNot',
      '''
      @if not cart.isEmpty {
        <p>You have @cart.itemCount items in your cart.</p>
      }
      @else {
        <p>Your cart is empty.</p>
      }
      ''');
  Add('codeForeach',
      '''
      <ul>
      @ForEach (var product in productList) {
          <li>@product.name - @product.price</li>
      }
      </ul>
      ''');
  Add('codeQuery', '<p>You searched for: @query.searchTerm</p>');
  Add('codePage', '<p>Current page is: @page.pagename</p>');
  Add('codeDataObject', 'WebStencilsProcessor1.AddVar(''user'', UserObject);');
  Add('codeDataMethod',
      '''
      WebStencilsProcessor1.AddVar('products',
        function: TObject
        begin
          Result := GetProductList;
        end);
      ''');
  Add('codeDataAttribute',
      '''
      type
        TMyDataModule = class(TDataModule)
          [WebStencilsVar('customers', false)]
          FDMemTable1: TFDMemTable;
          [WebStencilsVar('users')]
          FUsers: TUsers;
        end;

      WebStencilsProcessor1.AddModule(DataModule1);
      ''');
  Add('codeLayoutPage',
      '''
      @LayoutPage BaseTemplate
      <h1>Welcome to My Page</h1>
      <p>This is the content of my page.</p>
      ''');
  Add('codeRenderBody',
      '''
      <!DOCTYPE html>
      <html>
      <head>
          <title>My Website</title>
      </head>
      <body>
          <header>
              <!-- Common header content -->
          </header>

          <main>
              @RenderBody
          </main>

          <footer>
              <!-- Common footer content -->
          </footer>
      </body>
      </html>
      ''');
  Add('codeImport',
      '''
      @Import Sidebar.html

      @* Same behaviour as the previous one *@
      @Import Sidebar

      @* Use nested templates for better organization *@
      @Import folder/Sidebar
      ''');

  Add('codeHeaderBodyFooter',
      '''
      @Import Header.html

      <main>
          <!-- Page-specific content here -->
      </main>

      @Import Footer.html
      ''');

  Add('codeReusableComponents',
      '''
      <div class="product-list">
          @Import ProductList { @list = @ProductList }
      </div>

      <div class="tasks">
        @ForEach (var Task in Tasks.AllTasks) {
              @Import partials/tasks/item { @Task }
          }
      </div>
      ''');
end;

end.

File: Episode024/WebStencilsDemos/RADServerProject/C++/CodeExamplesU.cpp
#include <vcl.h>
#pragma hdrstop

#include "CodeExamplesU.h"
#pragma package(smart_init)

TCodeExample::TCodeExample(const String& AId, const String& ACode) {
    FId = AId;
    FCode = ACode;
}

TCodeExamples::TCodeExamples(TWebStencilsEngine* AWebStencilsEngine) {
    if (!AWebStencilsEngine) {
        throw Exception("WebStencilsEngine cannot be null");
    }
    
    try {
        FItems = new TStringList();
        FItems->OwnsObjects = true;
        FWebStencilsEngine = AWebStencilsEngine;
        
        // Initialize values first
        InitValues();
        
        // Then add them to WebStencils
        for (int i = 0; i < FItems->Count; i++) {
            TCodeExample* example = dynamic_cast<TCodeExample*>(FItems->Objects[i]);
            if (example) {
                try {
                    FWebStencilsEngine->AddVar(FItems->Strings[i], example);
                }
                catch (Exception& E) {
                    // Log the error but continue
                    printf("Error adding code example '%s': %s\n",
                           AnsiString(FItems->Strings[i]).c_str(),
                           AnsiString(E.Message).c_str());
                }
            }
        }
    }
    catch (...) {
        // Clean up if something goes wrong
        if (FItems) {
            delete FItems;
            FItems = nullptr;
        }
        throw;
    }
}

TCodeExamples::~TCodeExamples() {
    if (FItems) {
        delete FItems;
        FItems = nullptr;
    }
}

void TCodeExamples::Add(const String& AId, const String& ACode) {
    if (!FItems) {
        throw Exception("TStringList not initialized");
    }
    FItems->AddObject(AId, new TCodeExample(AId, ACode));
}

TCodeExample* TCodeExamples::Find(const String& AId) {
    if (!FItems) {
        return nullptr;
    }
    int index = FItems->IndexOf(AId);
    if (index >= 0) {
        return dynamic_cast<TCodeExample*>(FItems->Objects[index]);
    }
    return nullptr;
}

void TCodeExamples::InitValues() {
    if (!FItems) {
        throw Exception("TStringList not initialized");
    }
    
    Add("codeAtSymbol", "@object.value");
    Add("codeDotNotation",
        "<h2>User Profile</h2>\n"
        "<p>Name: @user.name</p>\n"
        "<p>Email: @user.email</p>");
    Add("codeComments",
        "@* This is a comment and will not appear in the output *@\n"
        "<p>This will appear in the output</p>");
    Add("codeIfAndElse",
        "@if user.isLoggedIn {\n"
        "    <p>Welcome, @user.name!</p>\n"
        "}\n"
        "@else {\n"
        "    <p>Please log in to continue.</p>\n"
        "}");
    Add("codeIfNot",
        "@if not cart.isEmpty {\n"
        "  <p>You have @cart.itemCount items in your cart.</p>\n"
        "}\n"
        "@else {\n"
        "  <p>Your cart is empty.</p>\n"
        "}");
    Add("codeForeach",
        "<ul>\n"
        "@ForEach (var product in productList) {\n"
        "    <li>@product.name - @product.price</li>\n"
        "}\n"
        "</ul>");
    Add("codeQuery", "<p>You searched for: @query.searchTerm</p>");
    Add("codePage", "<p>Current page is: @page.pagename</p>");
    Add("codeDataObject", "WebStencilsProcessor1->AddVar(\"user\", UserObject);");
    Add("codeDataMethod",
        "WebStencilsProcessor1.AddVar('products',\n"
        "  function: TObject\n"
        "  begin\n"
        "    Result := GetProductList;\n"
        "  end);");
    Add("codeDataAttribute",
        "type\n"
        "TMyDataModule = class(TDataModule)\n"
        "  [WebStencilsVar('customers', false)]\n"
        "  FDMemTable1: TFDMemTable;\n"
        "  [WebStencilsVar('users')]\n"
        "  FUsers: TUsers;\n"
        "end;\n"
        "\n"
        "WebStencilsProcessor1.AddModule(DataModule1);");

    Add("codeLayoutPage",
        "@LayoutPage BaseTemplate\n"
        "<h1>Welcome to My Page</h1>\n"
        "<p>This is the content of my page.</p>");

    Add("codeRenderBody",
        "<!DOCTYPE html>\n"
        "<html>\n"
        "<head>\n"
        "    <title>My Website</title>\n"
        "</head>\n"
        "<body>\n"
        "    <header>\n"
        "        <!-- Common header content -->\n"
        "    </header>\n"
        "    <main>\n"
        "        @RenderBody\n"
        "    </main>\n"
        "    <footer>\n"
        "        <!-- Common footer content -->\n"
        "    </footer>\n"
        "</body>\n"
        "</html>");

    Add("codeImport",
        "@Import Sidebar.html\n"
        "@* Same behaviour as the previous one *@\n"
        "@Import Sidebar\n"
        "@* Use nested templates for better organization *@\n"
        "@Import folder/Sidebar");

    Add("codeHeaderBodyFooter",
        "@Import Header.html\n"
        "<main>\n"
        "    <!-- Page-specific content here -->\n"
        "</main>\n"
        "@Import Footer.html");

    Add("codeReusableComponents",
        "<div class=\"product-list\">\n"
        "    @Import ProductList { @list = @ProductList }\n"
        "</div>\n"
        "<div class=\"tasks\">\n"
        "  @ForEach (var Task in Tasks.AllTasks) {\n"
        "        @Import partials/tasks/item { @Task }\n"
        "    }\n"
        "</div>");
}
